File: 124.cpp
struct Solution {
    int dfs(TreeNode *node, int &res){
        if (node == nullptr){
            return 0;
        }
        int lf = this->dfs(node->left, res);
        int rf = this->dfs(node->right, res);
        res = max(res, node->val + max(lf, 0) + max(rf, 0));
        return node->val + max(0, max(lf, rf));
    }
    inline int maxPathSum(TreeNode* root) {
        if (root == nullptr){
            return 0;
        }
        int res = root->val;
        this->dfs(root, res);
        return res;
    }
};

File: sol/124.py
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        res = root.val
        def dfs(node):
            if node is None:
                return 0
            lf = dfs(node.left)
            rf = dfs(node.right)
            nonlocal res
            res = max(res, node.val + max(lf, 0) + max(rf, 0))
            return node.val + max(lf, rf, 0)
        dfs(root)
        return res
